src/Api/ApiProvider.php

```diff
@@ -0,0 +1,217 @@
+ * {@see Aws\Utils::orFn}. This method accepts providers as arguments and
+ * returns a new function that will invoke each provider until a non-null value
+ * is returned.
+ *
+ *     $a = ApiProvider::filesystem(sys_get_temp_dir() . '/aws-beta-models');
+ *     $b = ApiProvider::manifest();
+ *
+ *     $c = Aws\Utils::orFn($a, $b);
+ *     $data = $c('api', 'betaservice', '2015-08-08'); // $a handles this.
+ *     $data = $c('api', 's3', '2006-03-01');          // $b handles this.
+ *     $data = $c('api', 'invalid', '2014-12-15');     // Neither handles this.
+ */
+class ApiProvider
+{
+    /**
+     * Resolves an API provider and ensures a non-null return value.
+     *
+     * @param callable $provider Provider function to invoke.
+     * @param string   $type     Type of data ('api', 'waiter', 'paginator').
+     * @param string   $service  Service name.
+     * @param string   $version  API version.
+     *
+     * @return array
+     * @throws UnresolvedApiException
+     */
+    public static function resolve(callable $provider, $type, $service, $version)
+    {
+        $result = $provider($type, $service, $version);
+        if (is_array($result)) {
+            return $result;
+        }
+
+        throw new UnresolvedApiException($service
+            ? "The {$service} service does not have API version: {$version}."
+            : "You must specify a valid service name to retrieve its API data."
+        );
+    }
+
+    /**
+     * Retrieves a list of valid versions for the specified service.
+     *
+     * @param string $service Service name
+     *
+     * @return array
+     */
+    public static function getServiceVersions($service)
+    {
+        $manifest = self::getManifestData();
+
+        if (!isset($manifest[$service])) {
+            return [];
+        }
+
+        return array_values(array_unique($manifest[$service]));
+    }
+
+    /**
+     * Default SDK API provider.
+     *
+     * This provider loads pre-built manifest data from the `data` directory.
+     *
+     * @return callable
+     */
+    public static function defaultProvider()
+    {
+        return self::manifest(self::getManifestData());
+    }
+
+    /**
+     * Loads API data after resolving the version to the latest, compatible,
+     * available version based on the provided manifest data.
+     *
+     * Manifest data is essentially an associative array of service names to
+     * associative arrays of API version aliases.
+     *
+     * [
+     *   ...
+     *   'ec2' => [
+     *     'latest' => '2014-10-01',
+     *     '2014-10-01' => '2014-10-01',
+     *     '2014-09-01' => '2014-10-01',
+     *     '2014-06-15' => '2014-10-01',
+     *     ...
+     *   ],
+     *   'ecs' => [...],
+     *   'elasticache' => [...],
+     *   ...
+     * ]
+     *
+     * @param array $manifest The API version manifest data.
+     *
+     * @return callable
+     */
+    public static function manifest(array $manifest, $dir = null)
```

Comment:
What would you think about making $dir required here, and the `defaultProvider` function passes in the relative directory pointing to the models shipped with the SDK?